stateresolutionv2.go

```diff
@@ -387,21 +380,39 @@ func (r *stateResolverV2) getFirstPowerLevelMainlineEvent(event *Event) (
 // the event is ignored and dropped. Returns two lists - the first contains the
 // accepted (authed) events and the second contains the rejected events.
 func (r *stateResolverV2) authAndApplyEvents(events []*Event) {
+	authEvents := NewAuthEvents(nil)
 	for _, event := range events {
-		// Check if the event is allowed based on the current partial state. If the
-		// event isn't allowed then simply ignore it and process the next one.
+		// Collect together the auth events. We'll start by collecting the
+		// auth event IDs from the event itself. These are our fallback if
+		// no partial state events are known.
+		authEvents.Clear()
+		for _, authEventID := range event.AuthEventIDs() {
+			if authEvent := r.authEventMap[authEventID]; authEvent != nil {
+				_ = authEvents.AddEvent(authEvent)
+			}
+		}
+
+		// Now layer on the partial state events that we do know. This should
+		// mean that we make forward progress.
+		for _, event := range []*Event{
+			r.resolvedCreate, r.resolvedJoinRules, r.resolvedPowerLevels,
+			r.resolvedMembers[event.Sender()], r.resolvedThirdPartyInvites[event.Sender()],
+		} {
+			if event != nil {
+				_ = authEvents.AddEvent(event)
+			}
+		}
+
+		// Check if the event is allowed based on the current partial state.
+		r.allower.update(&authEvents)
 		if err := r.allower.allowed(event); err != nil {
+			// The event was not allowed by the partial state, so skip it.
 			continue
```

Comment:
Partial state *and* fallback auth events presumably?

Reply:
That's right, yes, I'll update the comment.